libtenzir/builtins/operators/buffer.cpp

```diff
@@ -0,0 +1,415 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/detail/weak_run_delayed.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/tql2/plugin.hpp>
+#include <tenzir/uuid.hpp>
+
+#include <caf/typed_event_based_actor.hpp>
+
+#include <queue>
+
+namespace tenzir::plugins::buffer {
+
+namespace {
+
+using buffer_actor = caf::typed_actor<
+  // Write events into the buffer.
+  auto(atom::write, table_slice events)->caf::result<void>,
+  // Read events from the buffer.
+  auto(atom::read)->caf::result<table_slice>>;
+
+enum class buffer_policy {
+  block,
+  drop,
+};
+
+struct buffer_state {
+  static constexpr auto name = "buffer";
+
+  buffer_actor::pointer self = {};
+  located<uint64_t> capacity = {};
+  buffer_policy policy = {};
+  metric_handler metrics_handler = {};
+
+  uint64_t buffer_size = {};
+  std::queue<table_slice> buffer = {};
+  caf::typed_response_promise<table_slice> read_rp = {};
+
+  table_slice blocked_events = {};
+  caf::typed_response_promise<void> write_rp = {};
+
+  uint64_t num_dropped = {};
+
+  ~buffer_state() noexcept {
+    emit_metrics();
+    if (read_rp.pending()) {
+      read_rp.deliver(table_slice{});
+    }
+  }
+
+  auto write(table_slice events) -> caf::result<void> {
+    if (read_rp.pending()) {
+      read_rp.deliver(std::move(events));
+      return {};
+    }
+    if (buffer_size + events.rows() > capacity.inner) {
+      auto [lhs, rhs] = split(events, capacity.inner - buffer_size);
+      buffer_size += lhs.rows();
+      buffer.push(std::move(lhs));
+      switch (policy) {
+        case buffer_policy::drop: {
+          num_dropped += rhs.rows();
```

Comment:
I think we should consider emitting a warning here in addition to having the metrics.

Reply:
Agreed. Do you want to add the warning before I give this PR a test run?

Reply:
5e1a4264586